### ISSUE TITLE
Update to RequireJS string plugin

### ISSUE THREAD

**Issue by mbarlow12** [phetsims/chipper#715]
After some discussion, @zepumph , @jessegreenberg , and I think that implementing accessibility presents some new challenges with how we manage strings, namely in the volume of declarations currently required. For example, BASE has ~280 lines declaring strings, both for a11y and with `require`.  We’re guessing that as we implement a11y features for more complex sims, this number will likely grow, and it seems cumbersome and error-prone to ask devs to use that same pattern for strings.

Our hope is to implement some kind of string API update that would allow for more straightforward usage of the keys defined in the `strings.json` file (or files). Instead of:

```
var key1String = require( ‘string!SIM_NAME/sim-name.key1’ );
var key2String = require( ‘string!SIM_NAME/sim-name.key2’ );
var key3String = require( ‘string!SIM_NAME/sim-name.key3’ );
```
We’re hoping for something like:
```js
var SimNameStrings = require( ‘string!SIM_NAME’ ); // SimNameStrings.key1String -> ‘key1 value string'
```

Ideally, we could keep the original functionality in the event that only a single string is needed and to ensure that current usages still work. Additionally, the ability to define nested structures within the string file could prove useful:

```
{
    "a11y": {
        "buttonLabel": {
            "value": "the label for the button"
        },
        “anotherKey”: {
            “value”: “another accessible string"
        },
        ...
    },
    “pattern”: {
        "0Name1ValueUnitsVelocity": {
            "value": "{0} m/s"
          },
    },
    ...
}
```
could then be accessed with `SimNameStrings.a11y.buttonLabel` or `SimNameStrings.0Name1ValueUnitsVelocity`.

This would require some pretty significant updates to our RequireJS string plugin or creating a new `a11yString!` (or something similar) plugin, and we’re not sure exactly what the implementation would look like. We just wanted to address this in the early stages of a11y development.

Some initial pros:
- nesting within .json files
- all strings can be required in a single statement (and still utilize translation utility)
- backwards compatible with existing usages

Cons & additional work
- opaque process of assigning properties onto the `...Strings` object
- assigning ALL strings may not be desired
- harder to search accross project for usages of that string without naming collision possibilities
- unknown amount of work in the `chipper/js/require-js` file(s)
- new linting rules likely necessary

@jessegreenberg @zepumph is there anything I missed?

**Comment by zepumph** [comment]
Marking with an a11y label, but this would and could apply to all strings. I think you covered it all @mbarlow12.

**Comment by samreid** [comment]
Not 100% sure something like this is warranted.  Some thoughts:

0. Just to make it clear what is at stake--here is a snippet that loads strings in BalloonLocationDescriber.js (one of the files with a relatively large amount of strings for this sim).

```js
  var atWallString = BASEA11yStrings.atWall.value;
  var balloonStickingToString = BASEA11yStrings.balloonStickingTo.value;
  var balloonOnString = BASEA11yStrings.balloonOn.value;
  var balloonAtString = BASEA11yStrings.balloonAt.value;
  var balloonLocationAttractiveStatePatternString = BASEA11yStrings.balloonLocationAttractiveStatePattern.value;
  var releasedString = BASEA11yStrings.released.value;
  var initialMovementPatternString = BASEA11yStrings.initialMovementPattern.value;
  var continuousMovementWithLabelPatternString = BASEA11yStrings.continuousMovementWithLabelPattern.value;
  var twoBalloonInitialMovementPatternString = BASEA11yStrings.twoBalloonInitialMovementPattern.value;
  var twoBalloonNoChangeAndLocationPatternString = BASEA11yStrings.twoBalloonNoChangeAndLocationPattern.value;
  var twoBalloonNowDirectionPatternString = BASEA11yStrings.twoBalloonNowDirectionPattern.value;
  var extremelySlowlyString  = BASEA11yStrings.extremelySlowly.value;
  var verySlowlyString = BASEA11yStrings.verySlowly.value;
  var slowlyString = BASEA11yStrings.slowly.value;
  var quicklyString = BASEA11yStrings.quickly.value;
  var veryQuicklyString = BASEA11yStrings.veryQuickly.value;
  var upDraggingString = BASEA11yStrings.upDragging.value;
  var leftDraggingString = BASEA11yStrings.leftDragging.value;
  var downDraggingString = BASEA11yStrings.downDragging.value;
  var rightDraggingString = BASEA11yStrings.rightDragging.value;
  var upAndToTheRightDraggingString = BASEA11yStrings.upAndToTheRightDragging.value;
  var upAndToTheLeftDraggingString = BASEA11yStrings.upAndToTheLeftDragging.value;
  var downAndToTheRightDraggingString = BASEA11yStrings.downAndToTheRightDragging.value;
  var downAndToTheLeftDraggingString = BASEA11yStrings.downAndToTheLeftDragging.value;
  var upString = BASEA11yStrings.up.value;
  var leftString = BASEA11yStrings.left.value;
  var downString = BASEA11yStrings.down.value;
  var rightString = BASEA11yStrings.right.value;
  var upAndToTheRightString = BASEA11yStrings.upAndToTheRight.value;
  var upAndToTheLeftString = BASEA11yStrings.upAndToTheLeft.value;
  var downAndToTheRightString = BASEA11yStrings.downAndToTheRight.value;
  var downAndToTheLeftString = BASEA11yStrings.downAndToTheLeft.value;
  var atLeftEdgeString = BASEA11yStrings.atLeftEdge.value;
  var atTopString = BASEA11yStrings.atTop.value;
  var atBottomString = BASEA11yStrings.atBottom.value;
  var atRightEdgeString = BASEA11yStrings.atRightEdge.value;
  var onSweaterString = BASEA11yStrings.onSweater.value;
  var offSweaterString = BASEA11yStrings.offSweater.value;
  var balloonNewRegionPatternString = BASEA11yStrings.balloonNewRegionPattern.value;
  var closerToObjectPatternString = BASEA11yStrings.closerToObjectPattern.value;
  var sweaterString = BASEA11yStrings.sweater.value;
  var wallString = BASEA11yStrings.wall.value;
  var centerOfPlayAreaString = BASEA11yStrings.centerOfPlayArea.value;
  var rightEdgeOfPlayAreaString = BASEA11yStrings.rightEdgeOfPlayArea.value;
  var topEdgeOfPlayAreaString = BASEA11yStrings.topEdgeOfPlayArea.value;
  var bottomEdgeOfPlayAreaString = BASEA11yStrings.bottomEdgeOfPlayArea.value;
  var noChangeAndLocationPatternString = BASEA11yStrings.noChangeAndLocationPattern.value;
  var nearSweaterString = BASEA11yStrings.nearSweater.value;
  var balloonNearString = BASEA11yStrings.balloonNear.value;
  var locationAndInducedChargePatternString = BASEA11yStrings.locationAndInducedChargePattern.value;
  var singleStatementPatternString = BASEA11yStrings.singleStatementPattern.value;
  var interactionCueString = BASEA11yStrings.interactionCue.value;
  var balloonLabelWithAttractiveStatePatternString = BASEA11yStrings.balloonLabelWithAttractiveStatePattern.value;
  var balloonVeryCloseToString = BASEA11yStrings.balloonVeryCloseTo.value;
  var continuousMovementPatternString = BASEA11yStrings.continuousMovementPattern.value;
  var continuousMovementWithLandmarkPatternString = BASEA11yStrings.continuousMovementWithLandmarkPattern.value;
  var nowDirectionPatternString = BASEA11yStrings.nowDirectionPattern.value;
  var balloonLocationNoChangePatternString = BASEA11yStrings.balloonLocationNoChangePattern.value;
  var noChangeWithInducedChargePatternString = BASEA11yStrings.noChangeWithInducedChargePattern.value;
  var balloonLocationNearOtherPatternString = BASEA11yStrings.balloonLocationNearOtherPattern.value;
  var grabbedNonePatternString = BASEA11yStrings.grabbedNonePattern.value;
  var grabbedChargePatternString = BASEA11yStrings.grabbedChargePattern.value;
  var grabbedWithOtherChargePatternString = BASEA11yStrings.grabbedWithOtherChargePattern.value;
  var grabbedWithHelpPatternString = BASEA11yStrings.grabbedWithHelpPattern.value;
```

1. Balloons and Static Electricity is currently at 10619 lines of code.  280 lines of string declarations would be less than 3% of this number of lines.

2. This seems like it would require modifications to rosetta (and possibly Babel?) to make arbitrary nested json work.

3. Speaking of json, we may be able to use a requirejs JSON plugin (or use text! plugin combined with JSON.parse) instead of developing something ourselves from scratch.

4. Our existing system has several fail-safes, such as making sure that the requested string exists.  But if we used `BalloonsAndStaticElectricityStrings.buttonText` or `BalloonsAndStaticElectricityStrings['buttonText']`, we would not be able to detect mismatches between key names until `undefined` appears at runtime.

5. Wouldn't the same argument apply to all require statements?  For example, Balloons and Static Electricity has 356 lines of require statements  (even before converting the 240+ ally strings).  For instance, BalloonModel has:

```js
  var Range = require( 'DOT/Range' );
  var Vector2 = require( 'DOT/Vector2' );
  var Vector2IO = require( 'DOT/Vector2IO' );
```

Applying the arguments from strings to modules, wouldn't we also prefer to batch these statements as well?

```js
var DOT = require(`DOT`);
// ...
new DOT.Vector2()
```

6. The main hassle in (0) is having to manually type in and keep up to date the string names, but at least the lint tool keeps the guesswork out of naming the strings.  If we decide to stick with the single-string approach, we may benefit from developing our own micro-tools for automatically inserting string lines like that.  I've been using a script that automatically adds an `require` statement for a module, and it has come in very handy.

**Comment by pixelzoom** [comment]
Thoughts...

(1) Do something that works for all strings, and is not specific to a11y.

(2) Continue to support `require` of individual strings.  The approach that we've been using should continue to work.

(3) Create a new plugin that reads the string file into an object.  Let's not overload the existing `string!` plugin.  

(4) Read the string file once.  Subsequent calls should return a cached object.

(5) Rewrite the `string!` plugin to use this new plugin. That is, the first time you use `string!`, it calls the new plugin to create the object that contains all strings, then returns the one string that you asked for.

**Comment by samreid** [comment]
> (5) Rewrite the string! plugin to use this new plugin. That is, the first time you use string!, it calls the new plugin to create the object that contains all strings, then returns the one string that you asked for.

Our current system prunes or "treeshakes" out the unused strings.  Is that an important feature going forward?  What if scenery-phet or sun start accumulating numerous a11y strings?

> (4) Read the string file once. Subsequent calls should return a cached object.

Does the requirejs "text!" plugin automatically cache?  Also, here's an MIT json plugin if we want to look at it or leverage it: https://github.com/millermedeiros/requirejs-plugins/blob/master/src/json.js

**Comment by mbarlow12** [comment]
> (1) Do something that works for all strings, and is not specific to a11y.
> (2) Continue to support require of individual strings. The approach that we've been using should continue to work.

@pixelzoom I hadn't mentioned those ideas, but it was discussed in the a11y developer meeting. We all agreed that it would be ideal.

@samreid I'd also noticed that we weren't using the json plugin you linked; it could likely be really useful if we decide to move forward with something like this. Regarding:
>This seems like it would require modifications to rosetta (and possibly Babel?) to make arbitrary nested json work.
I don't think rosetta would need any major overhaul if we're treating all strings the same (a11y or no).

Finally, I think the distinction between strings and require statements in general is that (ideally) each require statement is handling a distinct feature, functionality, or visual element while the strings are all fetching content. Furthermore, the require statements aren't likely to grow in number all that much, while I can see the string statements ballooning as the sim interactions become more complex.

One other workaround we'd briefly considered was some kind of a grunt task or similar automated process that pulled all the strings into a single file using the requirejs toolchain and exposed it as an object, but that seemed like avoiding the real question of how we handle strings in general.

**Comment by jonathanolson** [comment]
The non-nested form of this would be pretty easy to support currently (I believe we pretty much create the desired JSON structure or something like it in memory).

Nesting will be somewhat complicated for rosetta (and figuring out what needs translation). Also it seems like if anyone ever names a string key `value` things will go horribly wrong. I'm not convinced of the advantages, namely that we've effectively been doing things like this with the current string keys (with separators).

I like the single require statement, BUT it's important to note this removes the ability to search for "unused" strings.

**Comment by mbarlow12** [comment]
Agreed on the difficulty of properly handling/translating nested JSON structures, but I don't want to get hung up on that as I see the real advantage as using a single require statement for the strings.

> BUT it's important to note this removes the ability to search for "unused" strings.

Would it be possible to handle this in a custom linting rule?

**Comment by mbarlow12** [comment]
From dev meeting 10/18/18:

- rosetta collects data on what strings are needed for translation
- potential issues around requiring strings from common components vs sim code
  - JO proposed `SceneryPhetStrings.getString( 'key' )` or similar API
- AP: may be useful to table until is becomes a larger issue
- continue discussion once JB returns

**Comment by samreid** [comment]
In discussion today, we concluded we will leave it as is.  @mbarlow12 points out that we may have to deal with this again when the strings have an order of magnitude more strings.  We agreed it is OK to close for now.
